engines/cegar_values.cpp

```diff
@@ -0,0 +1,301 @@
+    }
+
+    for (auto con : conc_ts_.constraints()) {
+      // NOTE: there should be better infrastructure for re-adding constraints
+      // currently have to avoid re-adding the next version
+      prover_ts_.add_constraint(va.visit(con.first), con.second);
+    }
+  } else {
+    Term trans = conc_ts_.trans();
+    static_cast<RelationalTransitionSystem &>(prover_ts_)
+        .set_behavior(va.visit(init), va.visit(trans));
+  }
+
+  assert(super::bad_);
+  super::bad_ = va.visit(super::bad_);
+  cegval_bad_ = to_cegval_solver_.transfer_term(super::bad_, BOOL);
+
+  // expecting to have had values to abstract
+  assert(prover_to_vals.size());
+  // copy over to the other solver
+  for (const auto & elem : prover_to_vals) {
+    // also make sure the abstract variables are frozen
+    prover_ts_.assign_next(elem.first, elem.first);
+    to_vals_[to_cegval_solver_.transfer_term(elem.first)] =
+        to_cegval_solver_.transfer_term(elem.second);
+  }
+}
+
+template <class Prover_T>
+bool CegarValues<Prover_T>::cegar_refine()
+{
+  size_t cex_length = super::witness_length();
+
+  // create bmc formula for abstract system
+  Term bmcform = cegval_un_.at_time(cegval_ts_.init(), 0);
+  for (size_t i = 0; i < cex_length; ++i) {
+    bmcform = cegval_solver_->make_term(
+        And, bmcform, cegval_un_.at_time(cegval_ts_.trans(), i));
+  }
+  bmcform = cegval_solver_->make_term(
+      And, bmcform, cegval_un_.at_time(cegval_bad_, cex_length));
+
+  cegval_solver_->push();
+  cegval_solver_->assert_formula(bmcform);
+
+  // TODO add lemmas to both cegval_ts_ and prover_ts_
+  TermVec assumps;
+  TermVec equalities;
+  for (const auto & elem : to_vals_) {
+    assert(cegval_ts_.is_curr_var(elem.first));
+    assert(elem.second->is_value());
+    Term lbl = cegval_labels_.at(elem.first);
+    assumps.push_back(lbl);
+    // since the variables are frozen, only need to add at time step 0
+    Term eqval = cegval_solver_->make_term(Equal, elem.first, elem.second);
+    equalities.push_back(eqval);
+    Term eqval0 = cegval_un_.at_time(eqval, 0);
+    Term imp = cegval_solver_->make_term(Implies, lbl, eqval0);
+    cegval_solver_->assert_formula(imp);
+  }
+  assert(assumps.size() == equalities.size());
+
+  Result r = cegval_solver_->check_sat_assuming(assumps);
+
+  // do refinement if needed
+  if (r.is_unsat()) {
+    UnorderedTermSet core;
+    cegval_solver_->get_unsat_core(core);
+    for (size_t i = 0; i < assumps.size(); ++i) {
+      if (core.find(assumps[i]) != core.end()) {
+        logger.log(2, "CegarValues adding refinement axiom {}", equalities[i]);
+        // need to refine both systems
+        cegval_ts_.add_constraint(equalities[i]);
+        Term transferred_eq =
+            from_cegval_solver_.transfer_term(equalities[i], BOOL);
+        // TODO this should be more modular
+        //      can't assume super::abs_ts_ is the right one to constrain
+        super::abs_ts_.add_constraint(transferred_eq);
```

Comment:
this is true for ic3ia and ceg-proph, but not for other engines.

We should specialize the refinement of the subprover ts, similar to what is done in the ceg-proph code.

Reply:
Good idea! Done in 182ddd9bb267fc0dac4027b638d0a593af15238a